### DISCUSSION TITLE
Sandboxing, containers, runtimes

### DISCUSSION THREAD

**Discussion by pfrazee** [atek-cloud/atek#9] [Core Development]
We need to settle on the runtime and how it will apply security & sandboxing policies. Currently just maintaining a list of options, please feel free to comment with your experience/thoughts. We should be mindful of ARM support and resource usage as RPIs are a big target for us.

- [libvirt](https://libvirt.org/) Toolkit for VMs
- [firecracker](https://firecracker-microvm.github.io/) "Light weight" qemu alternative
  - [Does support rpi/arm](https://news.ycombinator.com/item?id=21448766). [Is used by fly.io](https://fly.io/blog/sandboxing-and-workload-isolation/)
- [gvisor](https://gvisor.dev/) An application kernel for containers ("By providing each container with its own application kernel, gVisor limits the attack surface of the host.")
- [Kata Containers](https://katacontainers.io/). An OCI-compliant container runtime that executes containers within QEMU based virtual machines.
- Docker
- [podman](https://podman.io/) Dropin replacement for docker with no daemon

Related reading

- https://security.stackexchange.com/questions/169642/what-makes-docker-more-secure-than-vms-or-bare-metal
- https://cloud.google.com/blog/products/gcp/demystifying-container-vs-vm-based-security-security-in-plaintext
- https://dev.to/narasimha1997/building-a-secure-sandboxed-environment-for-executing-untrusted-code-7e8
- https://disaev.me/p/sandboxing-untrusted-code/
- https://news.ycombinator.com/item?id=25883253

**Comment by kaichaosun** [comment]
Is WASM something considered variable? Nowadays WebAssembly becomes more like a standard for safe and efficient sandboxing solution. It's not suitable for most existing self-hosted apps though.
Here is some dirty workflow,
* the client side in javascript and html lives in a webkit based browser, and make RPC to the WASM host.
* the WASM instance contains the backend logic like validate the messages is valid or not. Multiple WASM instances can be loaded and corresponds to different apps.
* The WSAM host then dispatch to the app backend lives as the wasm code.

**Reply by pfrazee** [reply]
WASM seems really promising. The ecosystem is still early, but we should dabble in it. A few thoughts:

- It seems unlikely that the wasm ecosystem will replace docker unless somebody creates a "docker/oci image to wasm" converter, which I haven't seen any signs of. The thinking around wasm containers is that they won't package entire system runtimes; they're a way to run *programs* rather than *containers*.
- We will probably need a way to run applications which can't convert into the wasm space, things like protocol and wallet daemons. (The hypercore daemon is an example of that.) For those, we will either need to support docker, or we'll just need to let those applications run without a sandbox.
- It's not 100% crazy to support unsandboxed services since that maximizes developer freedom (which is good) but it's a bit of a loaded gun so we have to think about it.

https://github.com/krustlet/krustlet is an interesting thing to look at. EDIT: interesting bunch of links here: https://medium.com/wasm/cloud-native-webassembly-90b5ed782ea2

**Comment by trasherdk** [comment]
WASM is useful bridging javascript to optimized libraries coded in other languages.

A good example is [monero-javascript](https://github.com/monero-ecosystem/monero-javascript)

* Access to `monerod` deamon RPC.
* Access to `wallet` RPC or In-Memory.
* Runs in Node.js applications.
* Runs in browser applications.
* Has detailed documentation.
